doc/DatabaseOperations.md

```diff
@@ -0,0 +1,255 @@
+[Bluebird](http://bluebirdjs.com/docs/why-promises.html) library, however,
+the operations described here use only `then()`, `resolve()`, and `reject()`
+functions, allowing the use of [Infusion's promise API](https://docs.fluidproject.org/infusion/development/promisesapi).
+
+## APIs
+
+The library consists of two fluid components, one that establishes a
+connection to the data base (host, port, etc.) and another that executes the
+data base requests.
+
+### `fluid.postgresdb.request`
+
+The request component is for initializing the connection with the database and
+providing a means to make requests of it.
+
+#### Component options
+
+These options are all set to `null` in the base grade.  Integrators need to set
+these as appropriate to their situation.
+
+| Option            | Type       | Description | Default |
+| ----------------- | ---------- | ----------- | ------- |
+| `databaseName`    | String     | Required. Name of the database containing the tables, e.g., `"fluid_prefsdb"` | `null` |
+| `host`            | String     | Required. The host for the requests, e.g., `"localhost"` | `null` |
+| `port`            | Integer    | Required. The port for the requests, e.g., `5432` | `null` |
+| `user`            | String     | Required. User administrator's name, e.g., `"admin"`| `null` |
+| `password`        | String     | Required for secure implemenations. User administrator's password | `null` |
+
+#### Component members
+
+| Member            | Type       | Description |
+| ----------------- | ---------- | ----------- |
+| `sequelize`       | Object     | Access to the Sequelize instance.
+
+When initialized, the request component establishes the connection with the
+database as per its options.  It is then available for database queries.
+However, an instance of the request component is a member of the
+`fluid.postgresdb.operations` component described immediately following.
+Clients use the operations component for most of the interactions with the
+database.
+
+### `fluid.postgresdb.operations`
+
+The majority of the database access operations is done using an instance of this
+component.  Its structure and API are documented here:
+
+#### Component members
+
+| Member             | Type       | Description |
+| ------------------ | ---------- | ----------- |
+| `tables`           | Object     | Hash of Sequelize table models, keyed by the table name.  For example, if there is a table in the database named `"users"`, then access to its model is via `tables["users"]` |
+| `modelDefinitions` | Object     | Hash of a table model definitions |
+
+#### Sub-components
+
+| Component         | Type                       | Description |
+| ----------------- | ----------                 | ----------- |
+| `request`         | `fluid.postgresdb.request` | An instance of the request component [documented above](#fluidpostgresdbrequest) |
+
+#### Operations API
+
+##### `createOneTable(tabelDef, deleteExisting)`
```

Comment:
The advantage of switching to PostgresSQL is the flexibility of SQL that can join any number of tables. For example, to verify an access token associated with an OAuth client and a key, the SQL looks like:
```
SELECT authorizations.accessToken 
FROM authorizations, keys, clientCredentials
WHERE keys.id = $inputKey
AND clientCredentials.oauth2ClientId = $inputOauth2ClientId
AND clientCredentials.oauth2ClientSecret = $inputOauth2ClientSecret
AND keys.id = authorizations.key
AND clientCredentials.oauth2ClientId = authorizations.clientCredentialId
```

When updating the preferences of a given key, the SQL could look like:
```
UPDATE prefsSafes
SET preferences = $preferences
FROM keys
WHERE keys.id = $inputKey
AND keys.prefsSafeId = prefsSafes.id
```

What I'm thinking is, instead of providing detailed APIs for creating/updating/inserting etc, it might be better to provide a simple API that takes in an input of a SQL statement and output the status of success/fail and the SQL result such as the retrieved records. This will save API users from translating a SQL to our API structure and save us from maintaining these APIs.

Reply:
I'm not sure what the standard process is for this, so please take my comments with a grain of salt. I just worry that passing in the SQL directly wouldn't be RESTful, if that's something we'd want to maintain. Unless you mapped SQL commands to HTTP methods. 

Also how does the security/permissions work. I'd imagine you'd want to lock down modifying the data, and for reading you may also want a system that can restrict portions of your preferences. e.g. only share my font size setting with my social media service.

Reply:
I think that Cindy is more talking about architectural API structure rather than necessarily anything that is exposed in a free-form API to the outside world. I agree that directly mapping to Postgres' relational model is preferable since I think we are all clear who won the 1980s battle between object-oriented thinkers and relational ones : P

Note that if we really did want to take this further and expose more free-form structures to the outside world, GraphQL is available as an adaptation of relational queries to an HTTP transport, but this would be unnecessarily heavyweight for our system.

I agree with Cindy that we should have a minimum number of architectural layers between the HTTP transport and the query level - people who work with our system will thank us for it later! (And I know that this is at odds with ways we've approached such problems in the past, and also with still-current ideas about abstraction of storage - but as our lengthy technology evaluation in the GPII demonstrated, one's choice of storage engine is actually in practice one of the least abstractable parts of a system, and it is perfectly valid for us at this point to say that our implementation is pretty closely tied to the choice of Postgres because we have devoted so much effort to analysing all the possible consequences of that choice in different contexts).

Reply:
> I'm not sure what the standard process is for this, so please take my comments with a grain of salt. I just worry that passing in the SQL directly wouldn't be RESTful, if that's something we'd want to maintain. Unless you mapped SQL commands to HTTP methods.

@jobara, Antranig has answered this part. These APIs are cloud side internal APIs that will be called by the server scripts behind the firewall. They are not exposed to the public.

> Also how does the security/permissions work. I'd imagine you'd want to lock down modifying the data, and for reading you may also want a system that can restrict portions of your preferences. e.g. only share my font size setting with my social media service.

The lock down when modifying the data is auto controlled by Postgres for concurrent data access. See [Postgres explicit locking](https://www.postgresql.org/docs/9.4/explicit-locking.html). In terms of accessing a portion of preferences, Postgres supports [JSON and JSONB data types](https://www.postgresql.org/docs/9.4/datatype-json.html) that a portion of JSON can be retrieved as `preferences -> fontSize`. This is one of the reasons that Postgres was chosen.

Reply:
@cindy sorry, my mistake about the level that this is referring to. @amb26 thanks for the clarification. I agree, it would simplify the process for communicating to the database. 

@cindyli the security question is probably irrelevant as I was thinking about this at a different level.

Reply:
@cindyli @amb26 

Oh.  I was trying to shield users from having to deal with raw SQL.  The idea behind the various API calls here was to capture frequent requests as a way of mitigating having to write complex SQL. Or even simple SQL for that matter.
And I realize that is a gamble trying to guess the most frequent or popular such calls.

That said, if it's better to have an interface that permits any SQL statement, I can live with that. Here are a few ways forward.

The code currently uses Sequelize to interface between the Postgres database and node.  Sequelize provides a [raw query function](https://sequelize.org/master/manual/raw-queries.html).  I'll check if it can handle your join example.  If so, that might be a way to proceed.

A second approach is to use node's [postgres](https://node-postgres.com/features/queries) module since it is  a JavaScript wrapper for Postgres.

A third approach is some other node postgres module that I'm unaware of.  If you can recommend a good one, I can look into it.

Reply:
@jobara At the risk of beating a dead horse -- @amb26  and @cindyli have already answered, and you've acknoweledged them -- users of the preferences server would not interact directly with the database.  They would make RESTful GET or POST requests such as `https://preferences.org/preferences/userid/retrieve?prefsSet=uio`, or something along those lines.  The details of that RESTful API are up for discussion in [GH-5](https://github.com/fluid-project/preferencesServer/issues/5).  Yes, the access needs to be secure and that would by handled by OAuth2 and/or login credentials.

Reply:
> A second approach is to use node's [postgres](https://node-postgres.com/features/queries) module since it is a JavaScript wrapper for Postgres.

I also looked at [node-postgres](https://node-postgres.com/), which is [the pg module](https://www.npmjs.com/package/pg) on npmjs. It looks promising to me.

Reply:
Node's postgres module ("pg") looks like an appropriate technology, especially since sequelize is advertising that it is short of maintainers. I don't think we will expose an interface which expects external users to write arbitrary SQL, but exposing intermediate operations which just abstractly create "a table" seems like a level of abstraction that they won't usually benefit from. We will write the SQL ourselves, and then package it behind an domain-appropriate API.

Reply:
node-postgres it is then, @cindyli @amb26.